File: 0015-3sum/0015-3sum.cpp
class Solution {
public:
    vector<vector<int>> threeSum(vector<int>& nums) {
        //TC: O(NlogN)
        sort(nums.begin(), nums.end());
        set<vector<int>> set;
        int n = nums.size();
        for(int i = 0; i < n; i++){
            int lo = i+1;
            int hi = n-1;
            while(lo < hi){
                if(nums[i] + nums[lo] + nums[hi] == 0){
                    set.insert({nums[i], nums[lo], nums[hi]});
                    lo++;
                }
                else if(nums[i] + nums[lo] + nums[hi] < 0){
                    lo++;
                }
                else{
                    hi--;
                }
            }
        }

        vector<vector<int>> ans(set.begin(), set.end());
        return ans;
    }
};

File: 0015-3sum/0015-3sum.py
class Solution(object):
    def threeSum(self, nums):
        """
        :type nums: List[int]
        :rtype: List[List[int]]
        """
        nums.sort()
        triplet_set = set()

        for i in range(len(nums)):
            j, k = i+1, len(nums)-1
            while j < k:
                sum = nums[i]+nums[j]+nums[k]
                if sum == 0:
                    triplet_set.add((nums[i], nums[j], nums[k]))
                    j += 1
                else:
                    if sum < 0:
                        j += 1
                    else:
                        k -= 1
        
        ans = [list(triplet) for triplet in triplet_set]

        return ans
        